### PR TITLE
adding 'select all' checkbox to multi select

### LINKED ISSUE #145: [FEATURE] - "Select All" option with toggle to clear
**Is your feature request related to a problem? Please describe.**

If there are a lot of options, it is a bit tedious to select them one by one if the user merely wants to select all of them.  In addition, it would be tedious to deselect all of the selected items.
**Describe the solution you'd like**

Please add a way to add functionality for a "Select All" option that would be clicked to select every item in the list, and when the user clicks on the Select All option again, it will clear the selections that have been made.  It would also be good if selecting all items manually automatically checked the "Select All" option (due to all items being selected).
**Describe alternatives you've considered**

mat-select-autocomplete-formcontrol does this, however your ngx-mat-select-search solution looks better visually.  I have attempted to do this myself, but I can only get "Select All" to work, but it takes a few seconds to work.  I cannot get Clear All to work, however.
**Additional context**

This is a similar implementation: https://stackblitz.com/edit/mat-select-autocomplete-formcontrol

### PR DESCRIPTION
**Pull request by blazewalker59**
uses form control to emit boolean to parent component parent selects all values in the filtered list

uses FormControl to listen for valuechanges on checkbox to pass to parent.
parent component then sets values in the filtered list to selected if boolean = true (checkbox is checked). if boolean is false, resets the selection list to empty array;

resolves #145

### CODE REVIEW

File: src/app/mat-select-search/mat-select-search.component.ts [hunk 1]
```diff
@@ -152,6 +152,10 @@ export class MatSelectSearchComponent implements OnInit, OnDestroy, AfterViewIni
   /** Adds 508 screen reader support for search box */
   @Input() ariaLabel = 'dropdown search';
```
Comment (macjohnny) [inline src/app/mat-select-search/mat-select-search.component.ts]:
```suggestion
@Input() showToggleAllCheckbox: boolean = false;
```

File: src/app/mat-select-search/mat-select-search.component.html [hunk 2]
```diff
@@ -6,6 +6,7 @@
       #innerSelectSearch
       class="mat-select-search-inner mat-typography mat-datepicker-content mat-tab-header"
       [ngClass]="{'mat-select-search-inner-multiple': matSelect.multiple}">
+  <mat-checkbox *ngIf="matSelect.multiple" color="primary" class="mat-select-all-checkbox" [formControl]="selectAllCheckbox"></mat-checkbox>
```
Comment (macjohnny) [inline src/app/mat-select-search/mat-select-search.component.html]: should be configurable with an input
```suggestion
  <mat-checkbox *ngIf="_isToggleAllCheckboxVisible()" color="primary" class="mat-select-all-checkbox" [formControl]="selectAllCheckbox"></mat-checkbox>
```

File: src/app/mat-select-search/mat-select-search.component.ts [hunk 3]
```diff
@@ -301,9 +305,21 @@ export class MatSelectSearchComponent implements OnInit, OnDestroy, AfterViewIni
...
 
+    // detect when select all checkbox changes
+    this.selectAllCheckbox.valueChanges
+      .pipe(takeUntil(this._onDestroy))
+      .subscribe(() => {
+        this.emitSelectAllBooleanToParent();
+      });
+
     this.initMultipleHandling();
   }
```
Comment (macjohnny) [inline src/app/mat-select-search/mat-select-search.component.ts]:
```suggestion
_isToggleAllCheckboxVisible(): boolean {
  return this.matSelect.multiple && this.showToggleAllCheckbox;
}
```

File: src/app/mat-select-search/mat-select-search.component.html [hunk 4]
```diff
@@ -6,6 +6,7 @@
       #innerSelectSearch
       class="mat-select-search-inner mat-typography mat-datepicker-content mat-tab-header"
       [ngClass]="{'mat-select-search-inner-multiple': matSelect.multiple}">
```
Comment (macjohnny) [inline src/app/mat-select-search/mat-select-search.component.html]:
```suggestion
      [ngClass]="{'mat-select-search-inner-multiple': matSelect.multiple, 'mat-select-search-inner-toggle-all': _isToggleAllCheckboxVisible()}">
```

File: src/app/mat-select-search/mat-select-search.component.scss [hunk 5]
```diff
@@ -33,10 +33,11 @@ $mat-option-height: 3em;
   -webkit-transform: translate3d(0,0,0);
 
   &.mat-select-search-inner-multiple {
-    width: 100%;
+    width: 80%;
```
Comment (macjohnny) [inline src/app/mat-select-search/mat-select-search.component.scss]:
```suggestion
    width: 100%;
    &.mat-select-search-inner-toggle-all {
      display: flex;
      align-items: center;
    }
```

File: src/app/mat-select-search/mat-select-search.component.scss [hunk 6]
```diff
@@ -33,10 +33,11 @@ $mat-option-height: 3em;
...
   }
   
   .mat-input-element {
+    width: 80%;
```
Comment (macjohnny) [inline src/app/mat-select-search/mat-select-search.component.scss]:
```suggestion
```

File: src/app/mat-select-search/mat-select-search.component.scss [hunk 7]
```diff
@@ -106,3 +107,8 @@ $mat-option-height: 3em;
     height: 2 * $mat-option-height;
   }
 }
+
+.mat-select-all-checkbox {
+  line-height: 50px;
```
Comment (macjohnny) [inline src/app/mat-select-search/mat-select-search.component.scss]:
```suggestion
```

File: src/app/mat-select-search/mat-select-search.component.ts [hunk 8]
```diff
@@ -152,6 +152,10 @@ export class MatSelectSearchComponent implements OnInit, OnDestroy, AfterViewIni
   /** Adds 508 screen reader support for search box */
   @Input() ariaLabel = 'dropdown search';
 
+  /** Output emitter to send to parent component with the select all boolean */
+  @Output() selectAllBooleanEmitter = new EventEmitter<boolean>();
```
Comment (macjohnny) [inline src/app/mat-select-search/mat-select-search.component.ts]:
```suggestion
  @Output() toggleSelectAll = new EventEmitter<boolean>();
```

File: src/app/mat-select-search/ngx-mat-select-search.module.ts [hunk 9]
```diff
@@ -5,19 +5,23 @@
...
 import { MatSelectSearchClearDirective } from './mat-select-search-clear.directive';
+import { FormsModule, ReactiveFormsModule } from '@angular/forms';
 
 export const MatSelectSearchVersion = '1.7.5';
 
 
 @NgModule({
   imports: [
     CommonModule,
+    FormsModule,
```
Comment (macjohnny) [inline src/app/mat-select-search/ngx-mat-select-search.module.ts]:
```suggestion
```

File: src/app/mat-select-search/ngx-mat-select-search.module.ts [hunk 10]
```diff
@@ -29,6 +33,9 @@ export const MatSelectSearchVersion = '1.7.5';
   exports: [
     MatSelectSearchComponent,
     MatSelectSearchClearDirective
+  ],
+  schemas: [
+    CUSTOM_ELEMENTS_SCHEMA
```
Comment (macjohnny) [inline src/app/mat-select-search/ngx-mat-select-search.module.ts]:
```suggestion
```

File: src/app/mat-select-search/mat-select-search.component.ts (hunk 3 above)
Comment (blazewalker59) [reply src/app/mat-select-search/mat-select-search.component.ts]: this subscription is needed to send the value of the checkbox to the parent component when it changes over time. I have implemented your suggested function of _isToggleAllCheckboxVisible() in my latest commit, but have kept the subscription calling the emit function.

File: src/app/examples/02-multiple-selection-example/multiple-selection-example.component.html [hunk 11]
```diff
@@ -2,8 +2,8 @@
 <h3>Multiple selection</h3>
 <p>
   <mat-form-field>
-    <mat-select [formControl]="bankMultiCtrl" placeholder="Banks" [multiple]="true" #multiSelect>
-      <ngx-mat-select-search [formControl]="bankMultiFilterCtrl"></ngx-mat-select-search>
+    <mat-select [formControl]="bankMultiCtrl" placeholder="Banks" [multiple]="true" #multiSelect disableOptionCentering>
+      <ngx-mat-select-search [showToggleAllCheckbox]="true" (toggleSelectAll)="toggleSelectAll($event)" [formControl]="bankMultiFilterCtrl"></ngx-mat-select-search>
```
Comment (macjohnny) [inline src/app/examples/02-multiple-selection-example/multiple-selection-example.component.html]: could you please duplicate the multiple-selection example, with and without toggle checkbox?

File: src/app/examples/02-multiple-selection-example/multiple-selection-example.component.ts [hunk 12]
```diff
@@ -57,6 +57,15 @@ export class MultipleSelectionExampleComponent implements OnInit, AfterViewInit,
     this._onDestroy.complete();
   }
 
+  toggleSelectAll(selectAllValue: boolean){
+    this.filteredBanksMulti.subscribe(val => {
```
Comment (macjohnny) [inline src/app/examples/02-multiple-selection-example/multiple-selection-example.component.ts]:
```suggestion
    this.filteredBanksMulti
      .pipe(takeUntil(this._onDestroy))
      .subscribe(val => {
```

File: src/app/mat-select-search/mat-select-search.component.scss [hunk 13]
```diff
@@ -34,9 +34,14 @@ $mat-option-height: 3em;
...
   &.mat-select-search-inner-multiple {
     width: 100%;
+    &.mat-select-search-inner-toggle-all {
+      display: flex;
+      align-items: center;
+    }
   }
   
   .mat-input-element {
+    width: 100%;
```
Comment (macjohnny) [inline src/app/mat-select-search/mat-select-search.component.scss]:
```suggestion
    flex-base: auto;
```

File: src/app/mat-select-search/ngx-mat-select-search.module.ts [hunk 14]
```diff
@@ -5,19 +5,22 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
-import { NgModule } from '@angular/core';
+import { NgModule, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
```
Comment (macjohnny) [inline src/app/mat-select-search/ngx-mat-select-search.module.ts]:
```suggestion
import { NgModule } from '@angular/core';
```

File: src/app/mat-select-search/ngx-mat-select-search.module.ts [hunk 15]
```diff
@@ -5,19 +5,22 @@
...
 import { MatSelectSearchComponent } from './mat-select-search.component';
-import {MatButtonModule, MatInputModule, MatIconModule, MatProgressSpinnerModule} from '@angular/material';
+import { MatButtonModule, MatInputModule, MatIconModule, MatProgressSpinnerModule, MatCheckboxModule} from '@angular/material';
```
Comment (macjohnny) [inline src/app/mat-select-search/ngx-mat-select-search.module.ts]:
```suggestion
import { MatButtonModule, MatInputModule, MatIconModule, MatProgressSpinnerModule, MatCheckboxModule } from '@angular/material';
```

File: src/app/examples/02-multiple-selection-example/multiple-selection-example.component.html (hunk 11 above)
Comment (blazewalker59) [reply src/app/examples/02-multiple-selection-example/multiple-selection-example.component.html]: within the same component? or just create a new component called multiple-selection-select-all-example ?

Comment (macjohnny) [reply src/app/examples/02-multiple-selection-example/multiple-selection-example.component.html]: please create a new component called multiple-selection-select-all-example.
this way we can test all features by checking the examples

Comment (blazewalker59) [reply src/app/examples/02-multiple-selection-example/multiple-selection-example.component.html]: it has been added in the latest commit as 06- multiple-selection-select-all-example

**Pull request merged**